pkg/monitor/monitor/util.go

```diff
@@ -0,0 +1,679 @@
+			},
+		},
+		RoleRef: rbac.RoleRef{
+			Kind:     "ClusterRole",
+			Name:     cr.Name,
+			APIGroup: "rbac.authorization.k8s.io",
+		},
+	}
+}
+
+func getMonitorDeployment(sa *core.ServiceAccount, config *core.ConfigMap, secret *core.Secret, monitor *v1alpha1.TidbMonitor, tc *v1alpha1.TidbCluster) *apps.Deployment {
+	deployment := getMonitorDeploymentSkeleton(sa, monitor)
+	initContainer := getMonitorInitContainer(monitor, tc)
+	deployment.Spec.Template.Spec.InitContainers = append(deployment.Spec.Template.Spec.InitContainers, initContainer)
+	prometheusContainer := getMonitorPrometheusContainer(monitor, tc)
+	reloaderContainer := getMonitorReloaderContainer(monitor, tc)
+	deployment.Spec.Template.Spec.Containers = append(deployment.Spec.Template.Spec.Containers, prometheusContainer, reloaderContainer)
+	if monitor.Spec.Grafana != nil {
+		grafanaContainer := getMonitorGrafanaContainer(secret, monitor, tc)
+		deployment.Spec.Template.Spec.Containers = append(deployment.Spec.Template.Spec.Containers, grafanaContainer)
+	}
+	volumes := getMonitorVolumes(config, monitor, tc)
+	deployment.Spec.Template.Spec.Volumes = volumes
+	return deployment
+}
+
+func getMonitorDeploymentSkeleton(sa *core.ServiceAccount, monitor *v1alpha1.TidbMonitor) *apps.Deployment {
+	monitorLabel := label.New().Instance(monitor.Name).Monitor().Labels()
+	replicas := int32(1)
+
+	deployment := &apps.Deployment{
+		ObjectMeta: meta.ObjectMeta{
+			Name:            getMonitorObjectName(monitor),
+			Namespace:       monitor.Namespace,
+			Labels:          monitorLabel,
+			OwnerReferences: []meta.OwnerReference{controller.GetTiDBMonitorOwnerRef(monitor)},
+			Annotations:     monitor.Spec.Annotations,
+		},
+		Spec: apps.DeploymentSpec{
+			Replicas: &replicas,
+			Strategy: apps.DeploymentStrategy{
+				Type:          apps.RecreateDeploymentStrategyType,
+				RollingUpdate: nil,
+			},
+			Selector: &meta.LabelSelector{
+				MatchLabels: map[string]string{
+					label.InstanceLabelKey:  monitor.Name,
+					label.ComponentLabelKey: label.TiDBMonitorVal,
+				},
+			},
+			Template: core.PodTemplateSpec{
+				ObjectMeta: meta.ObjectMeta{
+					Labels: map[string]string{
+						label.InstanceLabelKey:  monitor.Name,
+						label.ComponentLabelKey: label.TiDBMonitorVal,
+					},
+				},
+
+				Spec: core.PodSpec{
+					ServiceAccountName: sa.Name,
+					InitContainers:     []core.Container{},
+					Containers:         []core.Container{},
+					Volumes:            []core.Volume{},
+					Tolerations:        monitor.Spec.Tolerations,
+					NodeSelector:       monitor.Spec.NodeSelector,
+				},
+			},
+		},
+	}
+	return deployment
+}
+
+func getMonitorInitContainer(monitor *v1alpha1.TidbMonitor, tc *v1alpha1.TidbCluster) core.Container {
+	c := "mkdir -p /data/prometheus\nchmod 777 /data/prometheus\n/usr/bin/init.sh"
+	if monitor.Spec.Grafana != nil {
+		c = "mkdir -p /data/prometheus /data/grafana\nchmod 777 /data/prometheus /data/grafana\n/usr/bin/init.sh"
+	}
```

Comment:
Write multiple-line string in raw string literals is more readable

Reply:
updated.